docs-infra: dependency updates, strictness options and linting fixes

gkalpak opened the pull request:
This PR updates some more dependencies in `aio/` (including `@angular/cli@9.0.0-next.8` and `@angular/material@9.0.0-next.0`), enables some strict options (e.g. `fullTemplateTypeCheck`, some strict TS `compilerOptions`) and fixes some linting wanrings.

See individual commits for details.

##
~~This PR builds on top of #32923 (so the first few commits are from there).~~
_#32923 merged. Rebased on master._

gkalpak committed:
build(docs-infra): clean up and update dependencies to match latest CLI

This commit includes the following types of changes:
- Remove unused dependencies.
- Move dev dependencies from `devDependencies` to `dependencies` (and
  vice versa for production dependencies).
- Update `@types/*`.
- Update dependencies to more closely match the dependencies installed
  by the latest CLI for new apps.

Also, ensured that the latest version of `webdriver-manager` (v12.1.7)
was installed for `protractor`, which correctly installs a ChromeDriver
version that is compatible with the latest version of Chrome.

gkalpak committed:
build(docs-infra): enable more TypeScript strictness flags

gkalpak committed:
build(docs-infra): update @angular/material to 9.0.0-next.0

Upgrading to @angular/material 9.0.0-next.0 increases the bundle size
slightly (~1.3KB). This is a natural expectation of library
fixes/improvements.

gkalpak committed:
refactor(docs-infra): replace `mousewheel` event with `wheel` event

This commit replaces the non-standard, depecated [mousewheel][1] event
(which is not supported in Firefox and possibly other browsers too) with
the standard [wheel][2] event, which works on all supported browsers
according to [MDN's compatibility table][3].

I manually tested it in Chrome, Firefox, Edge and IE11.

[1]: https://developer.mozilla.org/en-US/docs/Web/API/Element/mousewheel_event
[2]: https://developer.mozilla.org/en-US/docs/Web/API/Element/wheel_event
[3]: https://developer.mozilla.org/en-US/docs/Web/API/Element/wheel_event#Browser_compatibility

mary-poppins commented:
You can preview 3e51961 at https://pr32980-3e51961.ngbuilds.io/.

gkalpak committed:
build(docs-infra): enable `fullTemplateTypeCheck` (and fix failures)

mary-poppins commented:
You can preview 0038e4a at https://pr32980-0038e4a.ngbuilds.io/.

mary-poppins commented:
You can preview 8a958f6 at https://pr32980-8a958f6.ngbuilds.io/.

mary-poppins commented:
You can preview 4166569 at https://pr32980-4166569.ngbuilds.io/.

mary-poppins commented:
You can preview 65990e3 at https://pr32980-65990e3.ngbuilds.io/.

gkalpak committed:
refactor(docs-infra): fix some linting warnings

mary-poppins commented:
You can preview 42cd089 at https://pr32980-42cd089.ngbuilds.io/.

gkalpak committed:
build(docs-infra): update payload sizes

There seems to be a ~3KB regression for IVy introduced in 4e35e348a.

gkalpak committed:
build(docs-infra): update @angular/cli to 9.0.0-next.8

The payload size increase in the ES5 bundles is (at least partially)
expected, due to fixing some down-leveling corner cases.

Related CLI issue: angular/angular-cli#15673

mary-poppins commented:
You can preview ea01027 at https://pr32980-ea01027.ngbuilds.io/.

mary-poppins commented:
You can preview 72623f5 at https://pr32980-72623f5.ngbuilds.io/.

gkalpak commented:
Marking as blocked on #32923, because that one has to be merged first.

filipesilva reviewed (approved):
LGTM for dev-infra

mary-poppins commented:
You can preview ec5afc2 at https://pr32980-ec5afc2.ngbuilds.io/.

Splaktar commented on aio/src/app/app.component.html:
```diff
@@ -57,7 +57,7 @@
 </mat-sidenav-container>
 
-<div *ngIf="hasFloatingToc" class="toc-container no-print" [style.max-height.px]="tocMaxHeight" (mousewheel)="restrainScrolling($event)">
+<div *ngIf="hasFloatingToc" class="toc-container no-print" [style.max-height.px]="tocMaxHeight" (wheel)="restrainScrolling($event)">
```
👍 I did this in AngularJS Material recently.

Splaktar reviewed (approved):
💛 it!

petebacondarwin commented:
The commit with message 

> build(docs-infra): clean up and update dependencies to match latest CLI

appears to do more than its message body suggests.

It looks like it is fixing some strictness/linting errors with the Firebase tests?

petebacondarwin commented on aio/src/app/custom-elements/api/api-list.component.html:
```diff
@@ -21,10 +21,10 @@
 
 <article class="api-list-container l-content-small docs-content">
-  <div *ngFor="let section of filteredSections | async" >
+  <div *ngFor="let section of (filteredSections | async)!" >
```
This is a bit worrying because `filteredSections` is of type `Observable<ApiSection[]>` but the `AsyncPipe` has the following type overload

```ts
  transform<T>(obj: Observable<T>|null|undefined): T|null;
```

which appears to convert the output of `async` to `ApiSection[]|null`.

But the `*ngFor` appears to only accept `ApiSection[]|Iterable<ApiSection>|undefined`.

Looking at the code I think that `ngFor` does actually accept `null` or `undefined`, so I think this type check is too tight - cc @alxhub

Splaktar replied:
That issue is fixed in PR https://github.com/angular/angular/pull/31371.

petebacondarwin replied:
Marvellous!

petebacondarwin replied:
@gkalpak should we have an issue to revert this when that PR lands?

gkalpak replied:
Created #33051 and assigned it to me.

petebacondarwin commented on aio/src/app/custom-elements/toc/toc.component.ts:
```diff
@@ -20,7 +20,7 @@ export class TocComponent implements OnInit, AfterViewInit, OnDestroy {
   @ViewChildren('tocItem') private items: QueryList<ElementRef>;
   private onDestroy = new Subject();
-  private primaryMax = 4;
+  primaryMax = 4;
```
+1 - used in the template

petebacondarwin commented on aio/src/app/custom-elements/toc/toc.component.spec.ts:
```diff
@@ -465,10 +465,11 @@ class TestScrollService {
     this.activeItemIndex.next(index);
-    if (asap.scheduled !== undefined) {
-      asap.flush();
+    if (asapScheduler.actions.length > 0) {
```
Was `scheduled` deprecated?

petebacondarwin reviewed (approved):
LGTM - with a few minor comments.
A bit worried about the template type check failures, which would be good to run past @alxhub at some point.

filipesilva commented:
You can remove the pinned Terser dep btw, a new version with a fix is available.

gkalpak commented on aio/src/app/layout/doc-viewer/doc-viewer.component.spec.ts:
```diff
@@ -47,7 +47,7 @@ describe('DocViewerComponent', () => {
     };
 
-    beforeEach(() => renderSpy = spyOn(docViewer, 'render').and.returnValue([null]));
+    beforeEach(() => renderSpy = spyOn(docViewer, 'render').and.returnValue(of(undefined)));
```
I prefer the current way, given that it is only used in two places (if it was used in more places it might be worth changing it). Having `.callThrough()` means you would have to go look at `TestDocViewerComponent` to see how it is implemented (e.g. "Does the observable emit synchronously? Does it emit automatically or do you have to manually trigger the emission (via a `Subject` or something?").

I don't feel strongly about it, though. Happy to change it, if you think it is better.

petebacondarwin replied:
Happy for the spy lines to be left as they are if you want. But I think we should fix the mock version to return the correct type in any case.

petebacondarwin replied:
(Although two test cases plus the mock definition breaks DRYness by a factor of three :-) )

gkalpak replied:
Fixed the return types and switched to `callThrough()`.

gkalpak replied:
Actually, I had to revert the change as it broke tests. It turns out that `TestDocViewerComponent` is only used for type-casting (in order to publicly expose protected methods of `DocViewerComponent` in tests). The implementation is still `DocViewerComponent`, so we need the spies to return mock values (and not `callThrough()`).

I added comments to the methods of `TestDocViewerComponent` to highlight that it is only used for type-casting. (I suspect you are not a fan of this pattern - and I am not a big fun either (nor do I know of any better one) - but changing it is outside the scope of this PR, so I am leaving it is as is for now :grin:)

You can take a look at the last two commits:
- https://github.com/angular/angular/pull/32980/commits/0858894bd: Revert the change and add comments.
- https://github.com/angular/angular/pull/32980/commits/76ce5f77c: Making the spies behave more similar to the actual implementation. (This doesn't make a difference in tests now. I just happened to notice it and decided to fix it, because it could potentially lead to hard-to-debug failures in the future.)

gkalpak commented on aio/src/app/custom-elements/api/api-list.component.ts:
```diff
@@ -112,28 +112,22 @@ export class ApiListComponent implements OnInit {
+      returnTrue : (item: ApiItem) => type === item.docType;
+
+    const items = section.items!.filter(item =>
+      matchesType(item) && matchesStatus(item) && matchesQuery(item));
```
I just wanted to move the function creation outside the loop (`.filter()`), so that we don't create new functions for _each item_. Other than that, any further optimization should be negligible. What would make them more readable for you? :grin:

Would that be better?

```ts
    const sectionNameMatches = !query || section.name.indexOf(query) !== -1;

    const matchesQuery = (item: ApiItem) =>
      sectionNameMatches || item.name.indexOf(query!) !== -1;
    const matchesStatus = (item: ApiItem) =>
      status === 'all' || status === item.stability || (status === 'security-risk' && item.securityRisk);
    const matchesType = (item: ApiItem) =>
      type === 'all' || type === item.docType;

    const items = section.items!.filter(item =>
      matchesType(item) && matchesStatus(item) && matchesQuery(item));
```

petebacondarwin replied:
Yep - that would be better IMHO. But it is subjective so 🤷‍♂

gkalpak commented on aio/scripts/_payload-limits.json:
```diff
@@ -4,8 +4,8 @@
         "runtime-es2015": 2996,
-        "main-es5": 501312,
-        "main-es2015": 456816,
+        "main-es5": 502599,
```
What do you mean? Find out what the extra code is?
(I assume this kind of small increase is expected as libraries add more features/fixes.)

petebacondarwin replied:
If that is the case, I think it is nice to write that in the commit message. E.g. "Upgrading to the latest Xxx library increases the bundle size slightly.  This is a natural expectation of library fixes." Just to be clear that this is not because of some regression in the bundler,  minifier, etc.

gkalpak commented on aio/tests/deployment/unit/tsconfig.json:
```diff
@@ -4,7 +4,7 @@
   },
   "include": [
-    "../shared/**/*.d.ts",
+    "../shared/**/*.ts",
```
Yes :smiley:

gkalpak commented on aio/tests/deployment/shared/helpers.ts:
```diff
@@ -1,3 +1,6 @@
+// tslint:disable-next-line: no-reference
+/// <reference path="./cjson.d.ts" />
```
`ts-node` (which we use these files with) ignore the `files` and `include` `tsconfig.json` options (see [here][1]). This was the simplest way I could make it work (adding `shared/` to the `typeRoots` didn't work for some reason and I didn't try `paths`, because triple-slash seemed simpler, less magical and worked :grin:).

[1]: https://github.com/TypeStrong/ts-node/tree/3cb340554#help-my-types-are-missing

gkalpak commented:
> You can remove the pinned Terser dep btw, a new version with a fix is available.

Thx, @filipesilva! I'll give that a try.

##
>> build(docs-infra): clean up and update dependencies to match latest CLI
>
> [...] appears to do more than its message body suggests.
> It looks like it is fixing some strictness/linting errors with the Firebase tests?

@petebacondarwin: You are right that the changes in [FirebaseGlob][1], [FirebaseRedirect][2] and [FirebaseRedirector][3] actually belong to the commit that turns on `noImplicitReturns`. I'll move them. Good spot 👍 

(The rest of the changes are necessary to account for the deps updates.)

[1]: https://github.com/angular/angular/commit/f9e0bfda1#diff-9a588366fd4fdb7a6699bad216ca2482
[2]: https://github.com/angular/angular/commit/f9e0bfda1#diff-b01cb3a05dbe9591e78a3b0ebc685aad
[3]: https://github.com/angular/angular/commit/f9e0bfda1#diff-a0afa4f4ebbc21f90daddb107d0a76c2

gkalpak committed:
fixup! build(docs-infra): clean up and update dependencies to match latest CLI

gkalpak committed:
fixup! build(docs-infra): enable more TypeScript strictness flags

gkalpak committed:
fixup! build(docs-infra): enable `fullTemplateTypeCheck` (and fix failures)

gkalpak committed:
build(docs-infra): unpin `terser` now that size regression is fixed

The latest terser version (4.3.8) includes a fix for the recent size
regression, so we can remove the pinning of transitive `terser`
dependencies to 4.3.2 (via `package.json > resolutions`).

mary-poppins commented:
You can preview ccf62bc at https://pr32980-ccf62bc.ngbuilds.io/.

petebacondarwin reviewed (approved):
Still LGTM!

mary-poppins commented:
You can preview 19dba64 at https://pr32980-19dba64.ngbuilds.io/.

gkalpak committed:
fixup! build(docs-infra): enable more TypeScript strictness flags

gkalpak committed:
test(docs-infra): ensure spy returns new observable every time

Previously, some spies in `DovViewerComponent` tests would return the
same `of(undefined)` observable for all invocations of the spy in a
test. While there is usually only one invocation per spy in each test,
this is not always the case. In case of multiple invocations within the
same test, subsequent calls would return an already completed
observable, which deviates from the actual behavior of the spied
function.

This commit fixes it by ensuring a fresh `of(undefined)` observable is
returned on each invocation.

mary-poppins commented:
You can preview 76ce5f7 at https://pr32980-76ce5f7.ngbuilds.io/.

gkalpak commented:
Thx for all the reviewing, @petebacondarwin. I had to make a couple of last changes (see https://github.com/angular/angular/pull/32980#discussion_r333021812).

Since the changes are minor and only affect tests, I am marking the PR for merging (but feel free to remove the `merge` label if you want to take another look first).

IgorMinar commented on aio/scripts/_payload-limits.json:
```diff
@@ -2,36 +2,36 @@
+        "runtime-es5": 2981,
+        "runtime-es2015": 2987,
+        "main-es5": 545197,
+        "main-es2015": 460505,
```
@kara and I suspect that this size regression is due to changes in Material. I don't think we can do much about that. :-(

gkalpak replied:
Can you elaborate a bit more? How come this change surfaced in this commit that does not update Material?

IgorMinar replied:
but oddly enough the view engine version has not regressed, which means that the regression might be caused by some other dependency change.

gkalpak replied:
What do you mean "has not regressed"? VE `main-es5` went from 484KB to 523KB 😕

alxhub commented on aio/src/app/custom-elements/api/api-list.component.html:
```diff
@@ -21,10 +21,10 @@
     <h2 *ngIf="section.items"><a [href]="section.path" [class.deprecated-api-item]="section.deprecated">{{section.title}}</a></h2>
     <ul class="api-list" *ngIf="section.items?.length">
-      <ng-container *ngFor="let item of section.items">
+      <ng-container *ngFor="let item of section.items!">
```
@petebacondarwin this is WAI for now.

`a?.b` and `a && a.b` have different semantics.

Longer term, it will be nice to our `?.` operator with safe navigation in TS 3.7, but I haven't looked into whether the semantics are a better match.
